src/Common/src/System/Security/Cryptography/RSASecurityTransforms.cs

```diff
@@ -225,36 +362,49 @@ public override byte[] SignHash(byte[] hash, HashAlgorithmName hashAlgorithm, RS
                     throw HashAlgorithmNameNullOrEmpty();
                 if (padding == null)
                     throw new ArgumentNullException(nameof(padding));
-                if (padding != RSASignaturePadding.Pkcs1)
-                    throw new CryptographicException(SR.Cryptography_InvalidPaddingMode);
 
-                SecKeyPair keys = GetKeys();
-
-                if (keys.PrivateKey == null)
+                if (padding == RSASignaturePadding.Pkcs1)
                 {
-                    throw new CryptographicException(SR.Cryptography_CSP_NoPrivateKey);
+                    SecKeyPair keys = GetKeys();
+
+                    if (keys.PrivateKey == null)
+                    {
+                        throw new CryptographicException(SR.Cryptography_CSP_NoPrivateKey);
+                    }
+
+                    int expectedSize;
+                    Interop.AppleCrypto.PAL_HashAlgorithm palAlgId =
+                        PalAlgorithmFromAlgorithmName(hashAlgorithm, out expectedSize);
+
+                    if (hash.Length != expectedSize)
+                    {
+                        // Windows: NTE_BAD_DATA ("Bad Data.")
+                        // OpenSSL: RSA_R_INVALID_MESSAGE_LENGTH ("invalid message length")
+                        throw new CryptographicException(
+                            SR.Format(
+                                SR.Cryptography_BadHashSize_ForAlgorithm,
+                                hash.Length,
+                                expectedSize,
+                                hashAlgorithm.Name));
+                    }
+
+                    return Interop.AppleCrypto.GenerateSignature(
+                        keys.PrivateKey,
+                        hash,
+                        palAlgId);
                 }
 
-                int expectedSize;
-                Interop.AppleCrypto.PAL_HashAlgorithm palAlgId =
-                    PalAlgorithmFromAlgorithmName(hashAlgorithm, out expectedSize);
+                // A signature will always be the keysize (in ceiling-bytes) in length.
+                byte[] output = new byte[(KeySize + 7) / 8];
```

Comment:
This math appears a lot... I know you'd proposed a public API for it, but short of that, how about an internal/private one?

Reply:
Added a utility method on RsaPaddingProcessor for now.